skbio/metadata/_interval.py

```diff
@@ -0,0 +1,274 @@
+
+        """
+        inv_md = Interval(_interval_metadata=self,
+                          intervals=intervals,
+                          boundaries=boundaries,
+                          metadata=metadata)
+
+        # Add directly to the tree.  So no need for _is_stale_tree
+        for loc in inv_md.intervals:
+            if loc is not None:
+                start, end = loc
+                self._intervals.add(start, end, inv_md)
+
+        self._metadata.append(inv_md)
+
+    def _rebuild_tree(self, intervals):
+        self._intervals = IntervalTree()
+        for f in intervals:
+            for inv in f.intervals:
+                start, end = inv
+                self._intervals.add(start, end, f)
+
+    def _query_interval(self, interval):
+        start, end = _polish_interval(interval)
+        invs = self._intervals.find(start, end)
+        return invs
+
+    def _query_attribute(self, key, value):
+        queries = []
+        for inv in self._metadata:
+            if inv[key] == value:
+                queries.append(inv)
+        return queries
+
+    def query(self, intervals=None, boundaries=None, metadata=None):
+        """ Looks up Interval objects with the intervals, boundaries and keywords.
+
+        Parameters
+        ----------
+        intervals : iterable of tuple of ints
+            A list of intervals associated with the Interval object
+        boundaries : iterable of tuple of bool
+            A list of boundaries associated with the Interval object
+        metadata : dict
+            A dictionary of key word attributes associated with the
+            Interval object
+
+        Returns
+        -------
+        list, Feature
+            A list of features satisfying the search criteria.
+
+        Examples
+        --------
+        >>> from skbio.metadata import IntervalMetadata
+        >>> im = IntervalMetadata()
+        >>> interval_metadata.add(intervals=[(0, 2), (4, 7)],
+                                  boundaries=None, metadata={'name': 'sagA'})
+        >>> interval_metadata.add(intervals=[(40, 70)],
+                                  boundaries=None, metadata={'name': 'sagB'})
+        >>> interval_metadata.query(intervals=[(1, 2)])
+        Interval(intervals=[(0, 2), (4, 7)], metadata={'name': 'sagA'})
+
+        Note
+        ----
+        There are two types of queries to perform
+        1. Query by interval
+        2. Query by key/val pair (i.e. gene=sagA)
+
+        """
+        if self._is_stale_tree:
+            self._rebuild_tree(self._metadata)
+            self._is_stale_tree = False
+
+        invs = []
+
+        # Find queries by interval
+        if intervals is not None:
+            for value in intervals:
+                invs += self._query_interval(value)
+
+        # Find queries by feature attribute
+        if metadata is not None:
+            for (key, value) in metadata.items():
+                invs += self._query_attribute(key, value)
+        return invs
```

Comment:
And yes, boundaries is not used. I did that on purpose, because I really don't know how this would be factored in.

Not sure how practical it is to make this a generator ...